### PR TITLE
Tool recommendation datatype fix

### PR DESCRIPTION
**Pull request by anuprulez**
Adds these features and fixes:
- datatype constraints for the predicted tools. Incompatible tools are filtered.
- the tools present in the tree (which appears after every tool run) are clickable. The clicked tool's definition opens in Galaxy.
- predicted tools tree UI fix.

### CODE REVIEW

**Comment by hexylena** [comment]
Hi @anuprulez your branch seems to overwrite some of the caching improvements I added. It took quite some time for me to rebase your previous changes on top of mine, would you mind updating your commits to be done off of the current version of the eu branch?

File: lib/galaxy/webapps/galaxy/api/workflows.py [hunk 1]
```diff
@@ -646,14 +642,13 @@ def get_tool_predictions(self, trans, topk=20, to_show=10, max_seq_len=25, paylo
...
-            except Exception:
+            except Exception as exception:
                 break
         # set the model weights
         self.loaded_model.set_weights(model_weights)
         tool_sequence = tool_sequence.split(",")
         tool_sequence = list(reversed(tool_sequence))
-
-        recommended_tools = self.__compute_tool_prediction(tool_sequence, topk, to_show, max_seq_len, self.model_data_dictionary, self.reverse_dictionary, self.compatible_tools, self.tool_weights_sorted, self.all_tools)
+        recommended_tools = self.__compute_tool_prediction(trans, tool_sequence, topk, to_show, max_seq_len)
```
Comment (hexylena) [inline lib/galaxy/webapps/galaxy/api/workflows.py:651]:
```suggestion
        recommended_tools = self.__compute_tool_prediction(tool_sequence, topk, to_show, max_seq_len, self.model_data_dictionary, self.reverse_dictionary, self.compatible_tools, self.tool_weights_sorted, self.all_tools)
```

Comment (hexylena) [reply lib/galaxy/webapps/galaxy/api/workflows.py:651]: No, ok, I misunderstood what changed.

**Comment by hexylena** [comment]
Ok new commits help :) I'll look more closely tomorrow and merge then

**Comment by anuprulez** [comment]
Yes, sure!
I took all your changes with local variables -> class variables. 
The variables are already attached to `self`. So, no need to pass them.

**Comment by anuprulez** [comment]
Can you look at it? @erasche 
thanks!

**Comment by hexylena** [comment]
Ok, got this merged in. Looks good!

**Pull request closed without merging**